pkg/clients/edgeconnect/client.go

```diff
@@ -0,0 +1,278 @@
+}
+
+// DetailsError represents details of errors
+type DetailsError struct {
+	ConstraintViolations ConstraintViolations `json:"constraintViolations"`
+	MissingScopes        []string             `json:"missingScopes,omitempty"`
+}
+
+// ConstraintViolations represents constraint violation errors
+type ConstraintViolations struct {
+	Message           string `json:"message"`
+	Path              string `json:"path,omitempty"`
+	ParameterLocation string `json:"parameterLocation,omitempty"`
+}
+
+// Error formats the server error code and message.
+func (e ServerError) Error() string {
+	if len(e.Message) == 0 && e.Code == 0 {
+		return "unknown server error"
+	}
+
+	return fmt.Sprintf("edgeconnect server error %d: %s", int64(e.Code), e.Message)
+}
+
+type serverErrorResponse struct {
+	ErrorMessage ServerError `json:"error"`
+}
+
+func CloseBodyAfterRequest(response *http.Response) {
+	if response != nil && response.Body != nil {
+		err := response.Body.Close()
+		if err != nil {
+			return
+		}
+	}
+}
+
+func (c *client) handleErrorResponseFromAPI(response []byte, statusCode int) error {
+	se := serverErrorResponse{}
+	if err := json.Unmarshal(response, &se); err != nil {
+		return errors.WithMessagef(err, "response error, can't unmarshal json response %d", statusCode)
+	}
+
+	return se.ErrorMessage
+}
+
+func (c *client) getServerResponseData(response *http.Response) ([]byte, error) {
+	responseData, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, errors.WithMessage(err, "error reading response")
+	}
+
+	if response.StatusCode != http.StatusOK &&
+		response.StatusCode != http.StatusCreated {
+		return responseData, c.handleErrorResponseFromAPI(responseData, response.StatusCode)
+	}
+
+	return responseData, nil
+}
+
+// GetEdgeConnect returns edge connect if it exists
+func (c *client) GetEdgeConnect(edgeConnectId string) (GetResponse, error) {
+	url := c.getEdgeConnectUrl(edgeConnectId)
+
+	resp, err := c.Client(c.ctx).Get(url)
+
+	if err != nil {
+		return GetResponse{}, err
+	}
+
+	defer CloseBodyAfterRequest(resp)
+
+	responseData, err := c.getServerResponseData(resp)
+	if err != nil {
+		return GetResponse{}, err
+	}
+
+	response := GetResponse{}
+
+	err = json.Unmarshal(responseData, &response)
+	if err != nil {
+		return GetResponse{}, err
+	}
+
+	return response, nil
+}
+
+// UpdateEdgeConnect updates existing edge connect hostPatterns and oauthClientId
+func (c *client) UpdateEdgeConnect(edgeConnectId, name string, hostPatterns []string, oauthClientId string) error {
+	url := c.getEdgeConnectUrl(edgeConnectId)
+
+	body := &Request{
+		Name:          name,
+		HostPatterns:  hostPatterns,
+		OauthClientId: oauthClientId,
+	}
+	payloadBuf := new(bytes.Buffer)
+	err := json.NewEncoder(payloadBuf).Encode(body)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPut, url, payloadBuf)
+
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.Client(c.ctx).Do(req)
+	defer CloseBodyAfterRequest(resp)
```

Comment:
For the request made in line 150, the "close body defer" is after the error check, here it's before. Would prefer to have it consistent.